File: src/Ledger.hs
module Ledger where

import Data.Time (Day)
import Data.List
import Money

data Account = Account String
  deriving (Eq, Show)

data ExternalEntity = ExternalEntity String
  deriving (Eq, Show)

data Balance = Balance {
  balanceDate :: Day,
  balanceAmount :: Money,
  balanceAccount :: Account
}
  deriving (Eq, Show)

data Event = Event {
  eventDate :: Day,
  eventAmount :: PositiveMoney, 
  eventTransaction :: Transaction
}
  deriving (Eq, Show)

type What = String
type FromAccount = Account
type ToAccount = Account

data Transaction = Income Account (Maybe ExternalEntity) (Maybe What)
                 | Expense Account (Maybe ExternalEntity) (Maybe What)
                 | Transfer FromAccount ToAccount
                 | StartOfDayBalance Account -- initial or balance override/correction
  deriving (Eq, Show)

applyTransaction :: Transaction -> PositiveMoney -> (Account -> Money) -> [(Account, Money)]
applyTransaction t positiveAmount amountBefore =
  case t of
    StartOfDayBalance account -> [update account $ const eventAmount]    
    Income account _ _ -> [update account (`plus` eventAmount)]
    Expense account _ _ -> [update account (`minus` eventAmount)]
    Transfer fromAccount toAccount -> [update fromAccount (`minus` eventAmount),
                                       update toAccount (`plus` eventAmount)]
  where
    eventAmount = toMoney positiveAmount
    update account calcNewAmount = (account, newAmount)
      where newAmount = calcNewAmount $ amountBefore account

applyEvent :: [Balance] -> Event -> [Balance]
applyEvent existingBalances (Event date amount transaction) =
  (filter isUnchanged existingBalances) ++ (map toBalance eventAmounts)
  where

    eventAmounts :: [(Account, Money)]
    eventAmounts = applyTransaction transaction amount amountBefore

    amountBefore :: Account -> Money
    amountBefore =
      (withDefault zero) . (fmap balanceAmount . existingBalance)
    
    existingBalance :: Account -> Maybe Balance
    existingBalance account =
      find ((==account) . balanceAccount) existingBalances

    eventAccounts = map fst eventAmounts
    existingAccounts = map balanceAccount existingBalances

    accountsUnchanged = existingAccounts \\ eventAccounts

    isUnchanged =
      null . (`lookup` eventAmounts) . balanceAccount
  
    toBalance (account, amount) =
      Balance date amount account

balanceHistory :: [Event] -> [Balance]
balanceHistory events =
  foldl' applyEvent [] $ sortOn eventDate events
  
withDefault :: a -> Maybe a -> a
withDefault a maybeA = maybe a id maybeA

    

File: src/main/scala/com/jm/budgetcalc/model/Ledger.scala
package com.jm.budgetcalc.model

import java.util.Date

case class Account(name: String)
case class ExternalEntity(name: String)

case class Balance(account: Account, date: Date, amount: Money)

case class LedgerEvent[T <: Transaction](date: Date, amount: PositiveMoney, transaction: T)

trait Transaction
case class SetAccountBalance(account: Account) extends Transaction
case class Income(account: Account, who: Option[ExternalEntity], what: Option[String]) extends Transaction
case class Expense(account: Account, who: Option[ExternalEntity], what: Option[String]) extends Transaction
case class Transfer(from: Account, to: Account) extends Transaction

object Transaction {

  type ChangedAccounts = Map[Account, Money]
  type AccountLookup = Account => Money
  type TransactionOp[T <: Transaction] = (T, PositiveMoney, AccountLookup) => ChangedAccounts

  def applyEvent[T <: Transaction](event: LedgerEvent[T], before: List[Balance])
                                  (implicit transactionOp: TransactionOp[T]): List[Balance] = {

    def amountBefore(account: Account): Money =
      before
        .find(_.account == account)
        .map(_.amount)
        .getOrElse(Money.zero)

    val changedBalances =
      transactionOp.apply(event.transaction, event.amount, amountBefore)
                   .map { case (account, amount) => Balance(account, event.date, amount) }

    val changedAccounts = changedBalances.map(_.account).toSet

    val unchangedBalances =
      before.filter(b => ! changedAccounts.contains(b.account))

    unchangedBalances ++ changedBalances
  }

  implicit def applySetAccountBalance(set: SetAccountBalance,
                                      amount: PositiveMoney,
                                      amountBefore: AccountLookup): ChangedAccounts =
      Map(set.account -> amount)

  implicit def applyIncome(income: Income,
                           amount: PositiveMoney,
                           amountBefore: AccountLookup): ChangedAccounts =
      Map(income.account -> (amountBefore(income.account) + amount))

  implicit def applyExpense(expense: Expense,
                            amount: PositiveMoney,
                            amountBefore: AccountLookup): ChangedAccounts =
      Map(expense.account -> (amountBefore(expense.account) - amount))

  implicit def applyTransfer(transfer: Transfer,
                             amount: PositiveMoney,
                             amountBefore: AccountLookup): ChangedAccounts =
      Map(
        transfer.from -> (amountBefore(transfer.from) - amount),
        transfer.to -> (amountBefore(transfer.to) + amount)
      )
}